blockchain/peer.go

```diff
@@ -0,0 +1,212 @@
+	logger log.Logger
+	ID     p2p.ID
+
+	Height                  int64                  // the peer reported height
+	NumPendingBlockRequests int32                  // number of requests still waiting for block responses
+	blocks                  map[int64]*types.Block // blocks received or expected to be received from this peer
+	blockResponseTimer      *time.Timer
+	recvMonitor             *flow.Monitor
+	params                  *bpPeerParams // parameters for timer and monitor
+
+	onErr func(err error, peerID p2p.ID) // function to call on error
+
+}
+
+// NewBPPeer creates a new peer.
+func NewBPPeer(
+	peerID p2p.ID, height int64, onErr func(err error, peerID p2p.ID), params *bpPeerParams) *bpPeer {
+
+	if params == nil {
+		params = bpPeerDefaultParams()
+	}
+	return &bpPeer{
+		ID:     peerID,
+		Height: height,
+		blocks: make(map[int64]*types.Block, maxRequestsPerPeer),
+		logger: log.NewNopLogger(),
+		onErr:  onErr,
+		params: params,
+	}
+}
+
+// String returns a string representation of a peer.
+func (peer *bpPeer) String() string {
+	return fmt.Sprintf("peer: %v height: %v pending: %v", peer.ID, peer.Height, peer.NumPendingBlockRequests)
+}
+
+// SetLogger sets the logger of the peer.
+func (peer *bpPeer) SetLogger(l log.Logger) {
+	peer.logger = l
+}
+
+// SetHeight sets the height of the peer.
+func (peer *bpPeer) SetHeight(height int64) {
+	peer.Height = height
+}
+
+// Cleanup performs cleanup of the peer, removes blocks, requests, stops timers and monitors.
+func (peer *bpPeer) Cleanup() {
+	if peer.blockResponseTimer != nil {
+		peer.blockResponseTimer.Stop()
+	}
+	if peer.NumPendingBlockRequests != 0 {
+		peer.logger.Info("peer with pending requests is being cleaned")
+	}
+	if int32(len(peer.blocks))-peer.NumPendingBlockRequests != 0 {
+		peer.logger.Info("peer with pending blocks is being cleaned")
+	}
+	for h := range peer.blocks {
+		delete(peer.blocks, h)
+	}
+	peer.NumPendingBlockRequests = 0
+	peer.recvMonitor = nil
+}
+
+// BlockAtHeight returns the block at a given height if available and a non-nil error otherwise.
+func (peer *bpPeer) BlockAtHeight(height int64) (*types.Block, error) {
+	block, ok := peer.blocks[height]
+	if !ok {
+		return nil, errMissingRequest
+	}
+	if block == nil {
+		return nil, errMissingBlock
+	}
+	return peer.blocks[height], nil
+}
+
+// RequestSent records that a request was sent, and starts the peer timer and monitor if needed.
+func (peer *bpPeer) RequestSent(height int64) {
+	peer.blocks[height] = nil
+
+	if peer.NumPendingBlockRequests == 0 {
+		peer.startMonitor()
+		peer.resetBlockResponseTimer()
+	}
+	peer.NumPendingBlockRequests++
+}
+
+// BlockReceived handles a block arrival at peer level. Block must be non-nil and recvSize a positive integer
+func (peer *bpPeer) BlockReceived(block *types.Block, recvSize int) error {
+	if block == nil || recvSize < 0 {
+		panic("bad parameters")
+	}
+	if peer.NumPendingBlockRequests == 0 {
+		panic("peer does not have pending requests")
```

Comment:
shouldn't we log and return errMissingRequest here too? otherwise any malicious peer will be able to crash our node by sending us a block

Reply:
Ouch, broken from last commit. The checks under should be before this.